pkg/controller/volume/attachdetach/reconciler/reconciler_test.go

```diff
@@ -741,39 +742,39 @@ func Test_Run_OneVolumeDetachFailNodeWithReadWriteOnce(t *testing.T) {
-	// The wait time is increasing exponentially from initial value of 0.5s (0.5, 1, 2, 4, ...).
-	// The test here waits for 100 Millisecond to make sure it is in exponential backoff period after
-	// the first detach operation. At this point, volumes status should not be updated
-	time.Sleep(100 * time.Millisecond)
-	verifyVolumeAttachedToNode(t, generatedVolumeName, nodeName1, cache.AttachStateAttached, asw)
-	verifyVolumeNoStatusUpdateNeeded(t, logger, generatedVolumeName, nodeName1, asw)
-
-	// Wait for 600ms to make sure second detach operation triggered. Again, The volume will be
-	// removed from being reported as attached on node status and then added back as attached.
-	// The volume will be in the list of attached volumes that need to be updated to node status.
-	time.Sleep(600 * time.Millisecond)
 	verifyVolumeAttachedToNode(t, generatedVolumeName, nodeName1, cache.AttachStateAttached, asw)
-	verifyVolumeReportedAsAttachedToNode(t, logger, generatedVolumeName, nodeName1, true, asw)
+	verifyVolumeReportedAsAttachedToNode(t, logger, generatedVolumeName, nodeName1, true, asw, volumeAttachedCheckTimeout)
+
+	// The below verifies check intermediate state and is flaky but useful
+	//
+	// // After the first detach fails, reconciler will wait for a period of time before retrying to detach.
+	// // The wait time is increasing exponentially from initial value of 0.5s (0.5, 1, 2, 4, ...).
+	// // The test here waits for 100 Millisecond to make sure it is in exponential backoff period after
+	// // the first detach operation. At this point, volumes status should not be updated
+	// time.Sleep(100 * time.Millisecond)
+	// verifyVolumeAttachedToNode(t, generatedVolumeName, nodeName1, cache.AttachStateAttached, asw)
+	// verifyVolumeNoStatusUpdateNeeded(t, logger, generatedVolumeName, nodeName1, asw)
+
+	// // Wait for 600ms to make sure second detach operation triggered. Again, The volume will be
+	// // removed from being reported as attached on node status and then added back as attached.
+	// // The volume will be in the list of attached volumes that need to be updated to node status.
+	// time.Sleep(600 * time.Millisecond)
+	// verifyVolumeAttachedToNode(t, generatedVolumeName, nodeName1, cache.AttachStateAttached, asw)
+	// verifyVolumeReportedAsAttachedToNode(t, logger, generatedVolumeName, nodeName1, true, asw, volumeAttachedCheckTimeout)
 
 	// Add a second pod which tries to attach the volume to the same node.
 	// After adding pod to the same node, detach will not be triggered any more.
 	generatedVolumeName, podAddErr = dsw.AddPod(types.UniquePodName(podName2), controllervolumetesting.NewPod(podName2, podName2), volumeSpec, nodeName1)
 	if podAddErr != nil {
 		t.Fatalf("AddPod failed. Expected: <no error> Actual: <%v>", podAddErr)
 	}
-	// Sleep 1s to verify no detach are triggered after second pod is added in the future.
-	time.Sleep(1000 * time.Millisecond)
+	// verify no detach are triggered after second pod is added in the future.
```

Comment:
IMO we should keep this `Sleep` call, because we are trying to ascertain effect of adding a pod to the reconciler and whether it results in any change. Without the `Sleep` what we are doing is, we are verifying immediate state and hence they are not the same thing.

Reply:
Updated.

Reply:
We are not "relying" on sleep. We are giving reconciler loop a chance to turn, so as we are testing code after reconciler loop has turned (at least in theory).